File: Juktaway/src/main/java/net/slashOmega/juktaway/fragment/main/tab/FavoritesFragment.java
package net.slashOmega.juktaway.fragment.main.tab;

import android.os.AsyncTask;
import android.view.View;

import java.util.ArrayList;

import net.slashOmega.juktaway.event.model.StreamingCreateFavoriteEvent;
import net.slashOmega.juktaway.event.model.StreamingUnFavoriteEvent;
import net.slashOmega.juktaway.model.AccessTokenManager;
import net.slashOmega.juktaway.model.FavRetweetManager;
import net.slashOmega.juktaway.model.Row;
import net.slashOmega.juktaway.model.TabManager;
import net.slashOmega.juktaway.model.TwitterManager;
import net.slashOmega.juktaway.settings.BasicSettings;
import twitter4j.Paging;
import twitter4j.ResponseList;
import twitter4j.Status;

/**
 * お気に入りタブ
 */
public class FavoritesFragment extends BaseFragment {

    /**
     * このタブを表す固有のID、ユーザーリストで正数を使うため負数を使う
     */
    public long getTabId() {
        return TabManager.FAVORITES_TAB_ID;
    }

    /**
     * このタブに表示するツイートの定義
     * @param row ストリーミングAPIから受け取った情報（ツイート＋ふぁぼ）
     *            CreateFavoriteEventをキャッチしている為、ふぁぼイベントを受け取ることが出来る
     * @return trueは表示しない、falseは表示する
     */
    @Override
    protected boolean isSkip(Row row) {
        return !row.isFavorite() || row.getSource().getId() != AccessTokenManager.getUserId();
    }

    @Override
    protected void taskExecute() {
        new FavoritesTask().execute();
    }

    private class FavoritesTask extends AsyncTask<Void, Void, ResponseList<Status>> {
        @Override
        protected ResponseList<twitter4j.Status> doInBackground(Void... params) {
            try {
                Paging paging = new Paging();
                if (mMaxId > 0 && !mReloading) {
                    paging.setMaxId(mMaxId - 1);
                    paging.setCount(BasicSettings.INSTANCE.getPageCount());
                }
                return TwitterManager.getTwitter().getFavorites(paging);
            } catch (OutOfMemoryError e) {
                return null;
            } catch (Exception e) {
                e.printStackTrace();
                return null;
            }
        }

        @Override
        protected void onPostExecute(ResponseList<twitter4j.Status> statuses) {
            mFooter.setVisibility(View.GONE);
            if (statuses == null || statuses.size() == 0) {
                mReloading = false;
                mPullToRefreshLayout.setRefreshComplete();
                mListView.setVisibility(View.VISIBLE);
                return;
            }
            if (mReloading) {
                clear();
                for (twitter4j.Status status : statuses) {
                    FavRetweetManager.INSTANCE.setFav(status.getId());
                    if (mMaxId <= 0L || mMaxId > status.getId()) {
                        mMaxId = status.getId();
                    }
                    mAdapter.add(Row.Companion.newStatus(status));
                }
                mReloading = false;
                mPullToRefreshLayout.setRefreshComplete();
            } else {
                for (twitter4j.Status status : statuses) {
                    FavRetweetManager.INSTANCE.setFav(status.getId());
                    if (mMaxId <= 0L || mMaxId > status.getId()) {
                        mMaxId = status.getId();
                    }
                    mAdapter.extensionAdd(Row.Companion.newStatus(status));
                }
                mAutoLoader = true;
                mListView.setVisibility(View.VISIBLE);
            }
        }
    }

    /**
     * ストリーミングAPIからふぁぼを受け取った時のイベント
     * @param event ふぁぼイベント
     */
    public void onEventMainThread(StreamingCreateFavoriteEvent event) {
        addStack(event.getRow());
    }

    /**
     * ストリーミングAPIからあんふぁぼイベントを受信
     * @param event ツイート
     */
    public void onEventMainThread(StreamingUnFavoriteEvent event) {
        ArrayList<Integer> removePositions = mAdapter.removeStatus(event.getStatus().getId());
        for (Integer removePosition : removePositions) {
            if (removePosition >= 0) {
                int visiblePosition = mListView.getFirstVisiblePosition();
                if (visiblePosition > removePosition) {
                    View view = mListView.getChildAt(0);
                    int y = view != null ? view.getTop() : 0;
                    mListView.setSelectionFromTop(visiblePosition - 1, y);
                    break;
                }
            }
        }
    }
}

File: Juktaway/src/main/java/net/slashOmega/juktaway/fragment/main/tab/FavoritesFragment.kt
package net.slashOmega.juktaway.fragment.main.tab

import android.view.View
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import net.slashOmega.juktaway.model.*
import net.slashOmega.juktaway.settings.BasicSettings
import net.slashOmega.juktaway.twitter.currentClient

class FavoritesFragment: BaseFragment() {
    override var tabId = TabManager.FAVORITES_TAB_ID

    override fun taskExecute() {
        GlobalScope.launch(Dispatchers.Main) {
            val statuses = runCatching {
                currentClient.favorite.run {
                    if (mMaxId > 0 && !mReloading) list(maxId = mMaxId, count = BasicSettings.pageCount)
                    else list(count = BasicSettings.pageCount)
                }.await()
            }.getOrNull()

            when {
                statuses.isNullOrEmpty() -> {
                    mReloading = false
                    mPullToRefreshLayout.setRefreshComplete()
                    mListView.visibility = View.VISIBLE
                }
                mReloading -> {
                    clear()
                    for (status in statuses) {
                        FavRetweetManager.setFav(status.id)
                        if (mMaxId <= 0L || mMaxId > status.id) mMaxId = status.id
                    }
                    mAdapter?.addAllFromStatusesSuspend(statuses)
                    mReloading = false
                    mPullToRefreshLayout.setRefreshComplete()
                }
                else -> {
                    for (status in statuses) {
                        FavRetweetManager.setFav(status.id)
                        if (mMaxId <= 0L || mMaxId > status.id) mMaxId = status.id
                    }
                    mAdapter?.extensionAddAllFromStatusesSuspend(statuses)
                    mAutoLoader = true
                    mListView.visibility = View.VISIBLE
                }
            }

            finishLoad()
        }
    }
}